metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ESGraphQueryDAO.java

```diff
@@ -2115,4 +2121,443 @@ private static boolean isRelationshipConnectedToInput(
 
     return false;
   }
+
+  /**
+   * Get lineage relationships up to the maximum number of relationships specified in the impact
+   * configuration. This method scrolls through results using search_after scroll until it reaches
+   * the limit or exhausts all results.
+   *
+   * @param opContext The operation context
+   * @param entityUrn The source entity URN
+   * @param lineageGraphFilters The lineage graph filters
+   * @param maxHops The maximum number of hops to traverse
+   * @return A LineageResponse containing all relationships up to the maxRelations limit
+   */
+  @WithSpan
+  public LineageResponse getImpactLineage(
+      @Nonnull final OperationContext opContext,
+      @Nonnull Urn entityUrn,
+      LineageGraphFilters lineageGraphFilters,
+      int maxHops) {
+
+    // Get the maxRelations limit from configuration
+    int maxRelations = config.getSearch().getGraph().getImpact().getMaxRelations();
+
+    Map<Urn, LineageRelationship> result = new HashMap<>();
+    long currentTime = System.currentTimeMillis();
+    long remainingTime = config.getSearch().getGraph().getTimeoutSeconds() * 1000;
+    long timeoutTime = currentTime + remainingTime;
+
+    // Do a Level-order BFS
```

Comment:
Any chance some of this can be extracted to a utility? Looks like a parallel impl of existing graph walk

Reply:
It is very similar however it deviates from the exploration logic.

The separation makes more sense to me because we're controlling the graph traversal in different ways. The UI exploration can explore in different directions in parallel, implements limited exploration at each level, requires special ordering at each level to prioritize certain edges (via), dynamic page sizes and probably some things I've forgotten.

I think it is difficult to understand the code when we're trying to solve both graph exploration logic for the UI and the one direction explore-all impact query. I feel that it will be easier to reason about the two methods of graph traversal separately. 

That said, there might be pieces still which I think we can combine. I'll take another pass as consolidation.